src/ipvs/ip_vs_proto_tcp.c

```diff
@@ -306,46 +306,129 @@ static void tcp_in_remove_ts(struct tcphdr *tcph)
 {
-    unsigned char *ptr;
-    int len, i;
-    uint32_t tcp_opt_len = af == AF_INET ? TCP_OLEN_IP4_ADDR : TCP_OLEN_IP6_ADDR;
+    unsigned char *ptr, *fast, *slow;
+    const unsigned char *l3hdr, *payload;
+    int i, optlen;
+    unsigned int pruned;
+    uint8_t opcode, opsize;
+    uint64_t opts_protected;
+    const uint8_t opts_maxlen[64] = {
+        [2] = 4,  [3] = 3,   [4] = 2,
+        [8] = 10, [30] = 40, [34] = 18
+    };
 
     ptr = (unsigned char *)(tcph + 1);
-    len = (tcph->doff << 2) - sizeof(struct tcphdr);
+    fast = slow = ptr;
+    optlen = (tcph->doff << 2) - sizeof(struct tcphdr);
+    payload = ptr + optlen;
 
-    while (len > 0) {
-        int opcode = *ptr++;
-        int opsize;
+    if (optlen < reqlen)    /* make no sense to do anything */
+        return 0;
 
+    while (optlen > 0) {
+        opcode = *ptr++;
         switch (opcode) {
         case TCP_OPT_EOL:
-            return;
+            goto fini;
         case TCP_OPT_NOP:
-            len--;
+            fast++;
+            optlen--;
             continue;
         default:
             opsize = *ptr++;
-            if (opsize < 2)    /* silly options */
-                return;
-            if (opsize > len)
-                return;    /* partial options */
-            if ((opcode == TCP_OPT_ADDR) && (opsize == tcp_opt_len)) {
-                for (i = 0; i < tcp_opt_len; i++) {
-                    *(ptr - 2 + i) = TCP_OPT_NOP;
+            if (opsize < 2)             /* silly options */
+                goto fini;
+            if (opsize > optlen)        /* partial options */
+                goto fini;
+            if (opcode == TCP_OPT_ADDR) {
+                fast += opsize;
+            } else {
+                for (i = 0; i < opsize; i++) {
+                    if (slow != fast)
+                        *slow = *fast;
+                    slow++;
+                    fast++;
                 }
-                /* DON'T RETURN
-                 * keep search other TCP_OPT_ADDR ,and clear them.
-                 * See https://github.com/iqiyi/dpvs/pull/925 for more detail. */
             }
 
             ptr += opsize - 2;
-            len -= opsize;
+            optlen -= opsize;
             break;
         }
     }
+
+fini:
+    pruned = payload - slow;
+    if (pruned < reqlen) {
+        /* further trim the options, the tcp functionality relies on unprotected
+         * options may get hurt, refer to:
+         *   https://www.iana.org/assignments/tcp-parameters/tcp-parameters.xhtml
+         *   #tcp-parameters-1
+         * */
+        ptr = slow;
+        slow = fast = (unsigned char *)(tcph + 1);
+        if (tcph->syn)
+            opts_protected = (1ULL << 2) | (1ULL << 3) | (1ULL << 4)    /* MSS, WS, SACKP */
+                | (1ULL << 8) | (1ULL << 30) | (1ULL << 34);            /* TS, MPTCP, TFO */
+        else
+            opts_protected = (1ULL << 8);    /* TS, drop SACK, MPTCP DSS/REMOVE_ADDR */
```

Comment:
TIMESTAMP was removed in SYN packet, should the packets contain TIMESTAMP option?

Reply:
It exists for similar reason as the above.